### PR TITLE
implement as_numer_denom

### PR DESCRIPTION
**Pull request by srajangarg**
Working on adding as_numer_denom from SymPy
- [x] Add
- [x] Mul
- [x] Rational
- [x] Power
- [x] Complex
- [x] Think of more classes where as_numer_denom can be simplified

### CODE REVIEW

**Comment by isuruf** [comment]
@srajangarg take a look at one of the Visitor classes like EvalRealDoubleVisitor to see how to do this.

**Comment by srajangarg** [comment]
@isuruf Okay, I read about visitor classes. What I can't figure out is how exactly will the function be called ? In principle `as_numer_denom()` should be a function of `Basic`, right? How exactly will this be called using a visitor? We'll call `accept()` with a custom made Visitor (on an instance) ? What about the two values we have to return? Sorry if my questions aren't very succint. How is it supposed to 'look'? (In comparision to something like this)
```
r1 = add(x, y);
r1->as_numer_denom(outArg(numer), outArg(denom));
```

**Comment by srajangarg** [comment]
@isuruf @Sumith1896 Nvm, got help from #734 thanks!

**Comment by srajangarg** [comment]
@isuruf  Does the structure look ok now?

File: symengine/basic.cpp [hunk 1]
```diff
@@ -47,5 +48,11 @@ RCP<const Basic> Basic::diff(const RCP<const Symbol> &x) const
     return Derivative::create(rcp_from_this(), {x});
 }
 
+void Basic::as_numer_denom(const Ptr<RCP<const Basic>> &numer, const Ptr<RCP<const Basic>> &denom) const
```
Comment (isuruf) [inline symengine/basic.cpp]: You can move this to `numer_denom.cpp`

File: symengine/basic.h [hunk 2]
```diff
@@ -158,6 +158,9 @@ class Basic : public EnableRCPFromThis<Basic> {
     //! Returns the list of arguments
     virtual vec_basic get_args() const = 0;
 
+    //! Returns the numerator/denominator form of the expression
+    virtual void as_numer_denom(const Ptr<RCP<const Basic>> &numer, const Ptr<RCP<const Basic>> &denom) const;
```
Comment (isuruf) [inline symengine/basic.h]: Add this method outside of the class like
`void as_numer_denom(const RCP<const Basic> &x, const Ptr<RCP<const Basic>> &numer, const Ptr<RCP<const Basic>> &denom)`

Comment (srajangarg) [inline symengine/basic.h]: @isuruf But doesn't calling the function like
`x->as_numer_denom(outArg(numer), outArg(denom))`, look much more intuitive than this
`as_numer_denom(x, outArg(numer), outArg(denom))`? What's the reason for not making it a member function? Just curious

Comment (isuruf) [inline symengine/basic.h]: We are trying to make everything non-member functions, but only keeping those that are already member functions. See https://github.com/symengine/symengine/pull/527#issue-94633819

File: symengine/tests/basic/test_as_numer_denom.cpp [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    RCP<const Basic> r1, num, den;
+
+    r1 = add(x, y);
+    as_numer_denom(r1, outArg(num), outArg(den));
+
+    REQUIRE(eq(*num, *r1));
+    REQUIRE(eq(*den, *one));
+
+    r1 = add(x, mul(y, pow(x, y)));
+    as_numer_denom(r1, outArg(num), outArg(den));
```
Comment (certik) [inline symengine/tests/basic/test_as_numer_denom.cpp]: There should be a test what the result in `num` and `den` is here.

**Comment by certik** [comment]
@srajangarg I think this looks very good and clean. I left a small comment above.

File: symengine/numer_denom.cpp [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+#include <symengine/mul.h>
+#include <symengine/pow.h>
+#include <symengine/functions.h>
+#include <symengine/constants.h>
+#include <symengine/visitor.h>
+
+namespace SymEngine {
+
+// need a .h file to fix this
+void as_numer_denom(const RCP<const Basic> &x, const Ptr<RCP<const Basic>> &numer, const Ptr<RCP<const Basic>> &denom);
```
Comment (isuruf) [inline symengine/numer_denom.cpp]: You can move this declaration to `basic.h`

**Comment by isuruf** [comment]
One other class that you can be separated to numerator and denominator is `Complex`

**Comment by srajangarg** [comment]
@certik  @isuruf 
The PR seems to be complete from my side. Do you have any more ideas?

File: symengine/basic.h [hunk 5]
```diff
@@ -220,6 +220,7 @@ bool is_a_sub(const Basic &b);
 RCP<const Basic> expand(const RCP<const Basic> &self);
 umap_short_basic series(const RCP<const Basic> &ex, const RCP<const Symbol> &var, unsigned int prec);
 umap_short_basic series_invfunc(const RCP<const Basic> &ex, const RCP<const Symbol> &var, unsigned int prec);
+void as_numer_denom(const RCP<const Basic> &x, const Ptr<RCP<const Basic>> &numer, const Ptr<RCP<const Basic>> &denom);
```
Comment (certik) [inline symengine/basic.h]: Why not move this into `numer_denom.h`? This algorithm doesn't need anything from the core, so I think it's cleaner if it lives separate from the core, isn't it?

File: symengine/numer_denom.cpp [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+                curr_num = add(mul(curr_num, divx), arg_num);
+                continue;
+            }
+
+            divx = div(curr_den, arg_den);
+            as_numer_denom(divx, outArg(divx_num), outArg(divx_den));
+            // if (eq(*divx_den, *one)) {
+            //     curr_num = add(mul(arg_num, divx), curr_num);
+            //     continue;
+            // } !! The below two 'general' statements cover this case
```
Comment (certik) [inline symengine/numer_denom.cpp]: Should the commented code be removed?

**Comment by certik** [comment]
@srajangarg I think that this looks very good, I only left two minor comments/questions. Thanks for the work.

File: symengine/numer_denom.cpp (hunk 6 above)
Comment (srajangarg) [inline symengine/numer_denom.cpp]: I think we should let it stay. Just explains what I've done. Always good for if changes are to be made in the future.

File: symengine/basic.h (hunk 5 above)
Comment (srajangarg) [inline symengine/basic.h]: Sure!

File: symengine/numer_denom.cpp (hunk 6 above)
Comment (certik) [inline symengine/numer_denom.cpp]: I don't understand what the commented code is doing, or supposed to do. Is it a simplified version of the two lines below? It doesn't seem to assign to `curr_den`.

Comment (srajangarg) [inline symengine/numer_denom.cpp]: It's the case where `the arg_den completely divides the curr_den` (opposite of the previous case). Now as `arg_den` completely divides `curr_den`, it doesn't have to be changed! After I check the two 'divide each other cases' I write the general version of 'multiplying denominators' (in an LCM sort of way) , by doing
`curr_den = mul(curr_den, divx_den);` and 
`curr_num = add(mul(curr_num, divx_den), mul(arg_num, divx_num));`

We see that in the case commented out, divx_den was anyways going to be `one` and `divx = divx_num`, which would be anyways handled by the two lines (written above) and give the same result as expected.

Comment (certik) [inline symengine/numer_denom.cpp]: Ah I see --- so the if statement is a special case, but it turns out that your general case already handles the special case. Is that right? Why don't you summarize what you just wrote in a small comment in the code to make this clear?

Comment (srajangarg) [inline symengine/numer_denom.cpp]: Done!

**Comment by certik** [comment]
Thanks. I restarted the failing build on Travis.

File: symengine/basic.h (hunk 5 above)
Comment (certik) [inline symengine/basic.h]: Let's just keep it as it is. We can always move things around later.

**Comment by certik** [comment]
That looks good to me and tests pass. @isuruf any other comments?

**Comment by Sumith1896** [comment]
+1 from my side too

**Comment by certik** [comment]
Thanks!

**Pull request merged**